Add a utility command to output an actor's MiniYaml tree

phrohdoh opened the pull request:
Possibly taking in a map path (to a .zip/.oramap/directory).

Usage: `orautil ts --actor-yaml World mods/ts/maps/arivruns/`

If you exclude the map path just the default ruleset will be used (files in the `Rules` section of the manifest).

GitHub won't let me reopen #10771.

pchote commented on OpenRA.Mods.Common/UtilityCommands/OutputActorMiniYamlCommand.cs:
```diff
@@ -0,0 +1,59 @@
+			RecursivePrintMiniYamlNodes(actorMiniYaml.Nodes, 1);
+		}
+
+		void RecursivePrintMiniYamlNodes(List<MiniYamlNode> nodes, int indent)
```
I'm pretty sure we already have code for this - we use it for saving miniyaml in e.g. the rule updater.

phrohdoh committed:
Add a utility command to output an actor's MiniYaml tree

Possibly taking in a path (to a .zip/.oramap/directory).

pchote commented on OpenRA.Mods.Common/UtilityCommands/OutputActorMiniYamlCommand.cs:
```diff
@@ -0,0 +1,59 @@
+			}
+		}
+
+		Dictionary<string, MiniYaml> LoadYamlRules(IReadOnlyFileSystem fs, Manifest manifest, MiniYaml additionalMiniYaml)
```
Why extract a function for a standard one-liner?

pchote replied:
You don't really need a dictionary either - just change the TryGetValue with a FirstOrDefault(n => n.Key ==)

Phrohdoh replied:
I prefer the dictionary approach.
There is nothing that prevents a property from sharing a name with an actor which will only lead to confusion if that happens.

But if you (or someone else) insists on this being done with LINQ that is a two minute change I can do.

pchote replied:
The query only applies to top-level nodes, so property names vs actor names doesn't come into this.

Phrohdoh commented:
Updated with most of @pchote's suggestions but I [kept the dictionary](https://github.com/OpenRA/OpenRA/pull/11031#discussion_r58310084).

This now outputs something similar to `# Rules for World (merged with map A River Runs Near It/042af00bc74f7a86536cb96cc89bc843a796c1c1):` as the first line of output before the actual rules.

Phrohdoh commented on OpenRA.Mods.Common/UtilityCommands/OutputActorMiniYamlCommand.cs:
```diff
@@ -0,0 +1,41 @@
+				map = new Map(modData, modData.ModFiles.OpenPackage(mapPath));
+
+			var fs = map ?? modData.DefaultFileSystem;
+			var tree = MiniYaml.Load(fs, modData.Manifest.Rules, map == null ? null : map.RuleDefinitions);
```
I'll rename this to `topLevelNodes` for clarity.

Phrohdoh commented on OpenRA.Mods.Common/UtilityCommands/OutputActorMiniYamlCommand.cs:
```diff
@@ -0,0 +1,41 @@
+			Game.ModData = modData;
+
+			var actorType = args[1];
+			var mapPath = args[2];
```
I think this is bogus.

If `Length == 2`:
[0] will be the command
[1] the actor type
[2] **exception**

if `Length == 3`:
[0] will be the command
[1] the actor type
[2] the map path

Will fix.

Phrohdoh commented:
Rebased and updated.

RoosterDragon commented on OpenRA.Mods.Common/UtilityCommands/OutputActorMiniYamlCommand.cs:
```diff
@@ -0,0 +1,40 @@
+
+			Map map = null;
+			if (args.Length == 3)
+				map = new Map(modData, modData.ModFiles.OpenPackage(args[2]));
```
You may wish to try-catch an invalid map path and exit gracefully.

RoosterDragon commented on OpenRA.Mods.Common/UtilityCommands/OutputActorMiniYamlCommand.cs:
```diff
@@ -0,0 +1,40 @@
+			var result = topLevelNodes.FirstOrDefault(n => n.Key == actorType);
+			if (result == null)
+			{
+				Console.WriteLine("Could not find actor '{0}'.", actorType);
```
Maybe include "(name is case-sensitive)" in the message?

RoosterDragon commented:
:+1:

obrakmann commented:
:+1: Looks good to me, other than the nits Rooster mentioned. Can you update this so it can be merged, please?

Phrohdoh commented:
Rebased and addressed both requests.

> tmba:openra thill $ orautil ts --actor-yaml E1 something/fake
Could not load map 'something/fake'.

Exited `2`.

> tmba:openra thill $ orautil ts --actor-yaml e1
Could not find actor 'e1' (name is case-sensitive).

Exited `1`

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5350a8f968587773a39c9dcf646ee46e8ca0269a)
